<html>
	<meta charset="UTF-8">
	<title>Protein Calculator v0.9.2b</title>
	<head>
		<link rel="stylesheet" type="text/css" href="protcalc.css">
	</head>
	<div class="menu">
		<a class="menu" href="http://pdb.org/pdb/" target="_blank">PDB</a>
		<a class="menu" href="http://www.ncbi.nlm.nih.gov/pubmed/" target="_blank">PubMed</a>
		<a class="menu" href="http://www.lifesci.dundee.ac.uk/groups/mike_ferguson/" target="_blank">Ferguson Group</a>
	</div> 
	<br>
	<body>
		<div class="ProtCalc">
		<h1>ProtCalc v0.9.2b</h1>
		<noscript>
		  You have JavaScript disabled or your browser does not support it. ProtCalc required JavaScript. 
		</noscript>
		<script src="protcalc.js"></script>
		<p>
		<FORM NAME="seqinput" ACTION="" METHOD="GET">
			<br>
			Input a FASTA formatted DNA or protein sequence:<BR>
			<textarea NAME="seqbox" value="Enter sequence", cols=80 rows=10></textarea><br><br>
			<!-- Or upload a .fasta file from your computer: 
			<input type=file size=20 name=seqfile><br> -->
			Sequence type:
			<input type="radio" name="FASTAtype" id="Protein" value=0 checked> Protein
			<input type="radio" name="FASTAtype" id="DNA" value=1> DNA<br><br>
			<input type="text" name="Absorbance" size="3"> Absorbance at 280nm (A)<br>
			<input type="text" name="l" size="3" value=1> Cuvette length (cm)<br><br>
			<input type="text" name="signFig" size="3" value=1>Significant figures<br><br>
			<input type="button" name="button" value="Calculate" onClick="parseInput(this.form)">
			<button type="button" onclick="blank(this.form)">Reset</button>

		</FORM></p>
		</div>
		<div class="table" id="translation">
			<div class="row">
				<div class="cell">Translated sequence:</div>
				<div class="prot" id="protSeq"></div><br>
			</div>
		</div>
		<div class="table" id="results">
			<h3>Results:</h3>
			<div class="row">
				<span class="cell">Protein:</span>
				<span class="cell" id="protName">None</span><br>
			</div>
			<div class="row">
				<span class="cell">Calculated molar mass:</span>
				<span class="cell" id="molMass">0</span><br>
			</div>
			<div class="row">
				<span class="cell">Calculated extinction coefficient:</span>
				<span class="cell" id="extCoeff">0</span><br>
				<span class="cell">(assuming all pairs of Cys residues form cystines)</span><br>
			</div>
			<div class="row">
				<span class="cell">Calculated concentration (&mu;mol/L):</span>
				<span class="cell" id="umolConc">0</span><br>
			</div>
			<div class="row">
				<span class="cell">Calculated concentration (mg/ml):</span>
				<span class="cell" id="mgMlConc">0</span><br>
			</div>
		</div>
		<br>
		
	</body>
	<p class="ref">
	Important Notes:
		<ul class="ref">
		<li>The &epsilon;(280) values uses "are quite reliable for proteins containing Trp residues, and less reliable for proteins that do not."(1)</li>
		</ul>
	</p><br>
	<p class="ref">References:
		<ul class="ref">
		<li>Pace CN, Vajdos F, Fee L, Grimsley G, Gray T. How to measure and predict the molar absorption coefficient of a protein. Protein Sci. 1995 Nov;4(11):2411-23. PubMed PMID: 8563639; PubMed Central PMCID: PMC2143013.</li>
		<li> Amino acid masses from <a href="http://web.expasy.org/findmod/findmod_masses.html#AA">ExPASY</a>. Nonstandard residue weights (B, Z, X) were handled <a href="http://web.expasy.org/compute_pi/pi_tool-doc.html">as documented in ExPASY's official tool (Comment #5)</a>, with data from <a href="http://web.expasy.org/docs/relnotes/relstat.html">ExPASY's statistics</a></li>
		</ul></p>
		<br><br><br>
</html>